columnflow/tasks/cms/external.py

```diff
@@ -0,0 +1,151 @@
+            if len(mc_profile) != len(data_profile):
+                raise Exception(
+                    f"the number of bins in the MC profile ({len(mc_profile)}) does not match that "
+                    f"of the data profile for the {shift} shift ({len(data_profile)})",
+                )
+
+            # store them
+            weights[shift] = [safe_div(d, m) for m, d in zip(mc_profile, data_profile)]
+            self.publish_message(f"computed pu weights for shift {shift}")
+
+        # save it
+        self.output().dump(weights, formatter="json")
+
+    @classmethod
+    def read_mc_profile_from_cfg(
+        cls,
+        pu_config_target: law.FileSystemTarget,
+    ) -> list[float]:
+        """
+        Takes a mc pileup configuration file stored in *pu_config_target*, parses its content and
+        return the pu profile as a list of float probabilities.
+        """
+        probs = []
+
+        # read non-empty lines
+        lines = map(lambda s: s.strip(), pu_config_target.load(formatter="text").split("\n"))
+
+        # loop through them and extract probabilities
+        found_prob_line = False
+        for line in lines:
+            # look for the start of the pu profile
+            if not found_prob_line:
+                if not line.startswith("mix.input.nbPileupEvents.probValue"):
+                    continue
+                found_prob_line = True
+                line = line.split("(", 1)[-1].strip()
+            # skip empty lines
+            if not line:
+                continue
+            # extract numbers
+            probs.extend(map(float, filter(bool, line.split(")", 1)[0].split(","))))
+            # look for the end of the pu profile
+            if ")" in line:
+                break
+
+        return cls.normalize_values(probs)
+
+    @classmethod
+    def read_data_profile_from_hist(
+        cls,
+        pu_hist_target: law.FileSystemTarget,
+    ) -> list[float]:
+        """
+        Takes the pileup profile in data preproducd by the lumi pog and stored in *pu_hist_target*,
+        builds the ratio to mc and returns the weights in a list.
+        """
+        hist_file = pu_hist_target.load(formatter="uproot")
+        probs = hist_file["pileup"].values().tolist()
+        return cls.normalize_values(probs)
+
+    @classmethod
+    def read_data_profile_from_pileupcalc(
+        cls,
+        pu_file_target: law.FileSystemTarget,
+        minbias_xs: float,
+    ) -> list[float]:
+        """
+        Takes the pileup profile in data read stored in *pu_file_target*, which should have been
+        produced when processing data, and a *minbias_mexs* value in mb (milli), builds the ratio to
+        mc and returns the weights in a list for 99 bins (recommended number).
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def normalize_values(cls, values: Sequence[float]) -> list[float]:
+        _sum = sum(values, 0.0)
+        return [value / _sum for value in values]
```

Comment:
I agree this is a very safe way to handle general `Sequence`s of floats. But the `for` loop isn't very vectorized, and the input values could also just be a numpy/awkward array, right?

Reply:
True. But anyway we deal with a maximum of 100 values here, so this should be too significant.